File: codeforces/SonyaAndQueries.cpp
#include <math.h>

#include <iostream>
#include <map>

using namespace std;

string convert_number_to_pattern(string n) {
    string answer = "";

    for (int i = 0; i < n.length(); i++) {
        int digit = int(n[i]);
        answer += to_string(digit % 2);
    }

    // remove leading zeros from answer
    while (answer[0] == '0') {
        answer = answer.substr(1);
    }
    return answer;
}

int main() {
    int t;
    cin >> t;

    map<string, int> temp_dict;
    for (int i = 0; i < t; i++) {
        char type;
        string value;
        cin >> type >> value;

        // remove leading zeros from value
        while (value[0] == '0') {
            value = value.substr(1);
        }

        if (type == '+') {
            temp_dict[convert_number_to_pattern(value)] += 1;
        } else if (type == '-') {
            temp_dict[convert_number_to_pattern(value)] -= 1;
        } else {
            cout << temp_dict[convert_number_to_pattern(value)] << endl;
        }
    }
}

// correct

File: codeforces/Sonya and Queries.py
def convert_number_to_pattern(n):
    answer = 0

    i = 0
    while n > 0:
        rem = n % 2
        answer += rem * (10**i)
        n = n // 10
        i += 1

    return answer


def main():
    temp_dict = {}
    for _ in range(int(input())):
        type, value = input().split()
        value = int(value)

        if type == "+":
            pattern = convert_number_to_pattern(value)
            if pattern in temp_dict:
                temp_dict[pattern] += 1
            else:
                temp_dict[pattern] = 1
        elif type == "-":
            pattern = convert_number_to_pattern(value)
            temp_dict[pattern] -= 1
        else:
            print(temp_dict.get(value, 0))


if __name__ == "__main__":
    main()

# tle because of python, else the solution is correct, and working in cpp.
